File: src/scala/Category.scala
/******************************************************************************

@LICENCE@
 
 *****************************************************************************/
 
package buyat.api.entities

/**
 * Represents a buy.at category.
 * @author buy.at
 * @version 1.0
 * @since 1.0
 * @param _categoryId Category ID.
 * @param _categoryName Category name.
 * @param _level Level (1 or 2).
 * @param _parentCategoryId ID of the parent category.
 * @param _parentCategoryName Name of the parent category.
 * @param _subcategories Subcategories of the category.
 */
class Category(_categoryId:Int, _categoryName:String, _level:Int, _parentCategoryId:Int,
                    _parentCategoryName:String, _subcategories:List[Category]) extends APIEntity {

  /**
   * Alternative constructor.
   * @param _categoryId Category ID.
   * @param _categoryName Category name.
   * @param _level Level (1 or 2).
   */
  def this(_categoryId:Int, _categoryName:String, _level:Int) = this(_categoryId, _categoryName, _level, -1, null, Nil)

  /**
   * ID of this category.
   */
  var categoryId:Int = _categoryId

  /**
   * Name of this category.
   */
  var categoryName:String = _categoryName

  /**
   * ID of parent category.
   * -1 if there is no parent category.
   */
  var parentCategoryId:Int = _parentCategoryId

  /**
   * Name of parent category.
   * Null if there is no parent category.
   */
  var parentCategoryName:String = _parentCategoryName

  /**
   * Level of this category.
   */
  var level = _level

  /**
   * List of subcategories.
   */
  var subcategories:List[Category] = _subcategories
}

File: src/csharp/Category.cs
/******************************************************************************

@LICENCE@
 
 *****************************************************************************/

using System;
using System.Collections.Generic;
using System.Text;

namespace BuyAtAPI
{

    public class Category
    {

        private BuyAtAPIClient apiClient;
        private BuyatAffiliateEntitiesCategory soapCategory;
        private List<Category> subcategories;

        /// <summary>
        /// Enum to define available levels
        /// </summary>
        public enum LevelType
        {
            /** Level 1 */
            Level1,
            /** Level 2 */
            Level2
        };

        /// <summary>
        /// Constructor, wrapping SOAP object
        /// </summary>
        public Category(BuyAtAPIClient apiClient, BuyatAffiliateEntitiesCategory soapCategory)
        {
            this.apiClient = apiClient;
            this.soapCategory = soapCategory;
        }

        /// <summary>
        /// The category ID
        /// </summary>
        public int CategoryID
        {
            get
            {
                try
                {
                    return Convert.ToInt32(soapCategory.category.category_id);
                }
                catch (FormatException)
                {
                    return -1;
                }
            }
        }

        /// <summary>
        /// The category's level
        /// </summary>
        public LevelType Level
        {
            get
            {
                if (soapCategory.category.level == "1")
                {
                    return LevelType.Level1;
                }
                else
                {
                    return LevelType.Level2;
                }
            }
        }

        /// <summary>
        /// The category name
        /// </summary>
        public string CategoryName
        {
            get
            {
                return soapCategory.category.category_name;
            }
        }


        /// <summary>
        /// The parent category ID, if this is a level 2 category
        /// </summary>
        public int ParentCategoryID
        {
            get
            {
                try
                {
                    return Convert.ToInt32(soapCategory.category.parent_category_id);
                }
                catch (FormatException)
                {
                    return -1;
                }
            }
        }

        /// <summary>
        /// The parent category name, if this is a level 2 category
        /// </summary>
        public string ParentCategoryName
        {
            get
            {
                return soapCategory.category.parent_category_name;
            }
        }

        /// <summary>
        /// The subcategories of this category, if it is a level 1 category
        /// </summary>
        public List<Category> Subcategories
        {
            get
            {
                if (this.subcategories == null && this.Level == LevelType.Level1)
                {
                    this.subcategories = this.apiClient.ListLevel2Categories(this.CategoryID);
                }
                return this.subcategories;
            }
        }

    }
}
